fcd/pass_argrec_remill.cpp

```diff
@@ -0,0 +1,370 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/InstIterator.h>
+
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "remill/BC/ABI.h"
+#include "remill/BC/Util.h"
+
+#include "fcd/pass_argrec_remill.h"
+
+namespace fcd {
+
+namespace {
+
+static const char *sPrefix;
+
+static std::unordered_set<const char *> RegisterAliasSet(const char *reg) {
+  static const char *aliases[][9] = {{"AL", "AH", "AX", "EAX", "RAX", nullptr},
+                                     {"CL", "CH", "CX", "ECX", "RCX", nullptr},
+                                     {"DL", "DH", "DX", "EDX", "RDX", nullptr},
+                                     {"SIL", "SI", "ESI", "RSI", nullptr},
+                                     {"DIL", "DI", "EDI", "RDI", nullptr},
+                                     {"SPL", "SP", "ESP", "RSP", nullptr},
+                                     {"R8B", "R8W", "R8D", "R8", nullptr},
+                                     {"R9B", "R9W", "R9D", "R9", nullptr},
+                                     {nullptr}};
+
+  static std::unordered_map<const char *, const char **> resolver;
+  if (resolver.empty()) {
+    for (unsigned i = 0; aliases[i][0] != nullptr; ++i) {
+      for (unsigned j = 0; aliases[i][j] != nullptr; ++j) {
+        resolver[aliases[i][j]] = aliases[i];
+      }
+    }
+  }
+
+  std::unordered_set<const char *> result({reg});
+  auto iter = resolver.find(reg);
+  if (iter != resolver.end()) {
+    for (unsigned i = 0; iter->second[i] != nullptr; ++i) {
+      result.insert(iter->second[i]);
+    }
+  }
+
+  return result;
+}
+
+static std::unordered_set<llvm::User *> UsersOfVar(llvm::Function *func,
+                                                   llvm::Value *var) {
+  std::unordered_set<llvm::User *> result;
+  if (var->hasNUsesOrMore(2)) {
+    for (auto var_user : var->users()) {
+      if (auto addr = llvm::dyn_cast<llvm::LoadInst>(var_user)) {
+        for (auto addr_user : addr->users()) {
+          result.insert(addr_user);
+        }
+      }
+    }
+  }
+  return result;
+}
+
+template <typename T>
+static llvm::User *FirstUserOfVar(llvm::Function *func, llvm::Value *var,
+                                  T &instruction_list) {
+  auto users = UsersOfVar(func, var);
+  if (!users.empty()) {
+    for (auto &inst : instruction_list) {
+      if (users.count(&inst) > 0) {
+        return &inst;
+      }
+    }
+  }
+  return nullptr;
+}
+
+template <typename T>
+static std::pair<llvm::User *, const char *> FirstUserOfReg(
+    llvm::Function *func, const char *reg, T &instruction_list) {
+  std::unordered_map<llvm::User *, const char *> users;
+  for (auto alias : RegisterAliasSet(reg)) {
+    auto var = remill::FindVarInFunction(func, alias);
```

Comment:
It's run before optimizations, precisely because the optimizations rename the variables and generally identifying registers post-opt is hard. And I definitely agree that a real alias analysis would be best. I'll make an issue for this for future development.

Reply:
Addressed in issue #17